### PR TITLE
support for static-route yang model

### PR DESCRIPTION
**Pull request by UmaMaven**
Signed-off-by: Uma Ramanathan [uma.ramanathan@mavenir.com](mailto:uma.ramanathan@mavenir.com)

Why I did it
Support yang model for static route

How I did it
Added a yang file for static route configuration

How to verify it
Add static routes with gateway/interface/blackhole as nexthop with VRF capability

https://github.com/sonic-net/sonic-buildimage/issues/11498

### CODE REVIEW

**Comment by UmaMaven** [comment]
@venkatmahalingam @prsunny @ganglyu  
Hi All, 

This issue had to be created to due to a few problem I had face while trying to commit 
https://github.com/sonic-net/sonic-buildimage/pull/11499
There in 11499, while I was trying to do make fresh pull request (pulling to latest), the changes were automatically committed with 0 files. 

This PR is another attempt in committing the changes for issue #11498. Kindly approve.
Right now, LGTM runs are cancelled by Shilong Liu @liushilongbuaa - Could someone let me know the reason?

Thanks,
Uma

**Review by ganglyu — APPROVED** [review]

**Review by venkatmahalingam — APPROVED** [review]

**Comment by liushilongbuaa** [comment]
> @venkatmahalingam @prsunny @ganglyu Hi All,
> 
> This issue had to be created to due to a few problems I had while trying to commit #11499 There in 11499, while I was trying to do make fresh pull request (pulling to latest), the changes were automatically committed with 0 files.
> 
> This PR is another attempt in committing the changes for issue #11498. Kindly approve. Right now, LGTM runs are cancelled by Shilong Liu @liushilongbuaa - Could someone let me know the reason?
> 
> Thanks, Uma

LGTM success 5days ago. But there is unit test error when building sonic-config-engine.

**Comment by ganglyu** [comment]
@UmaMaven 
Please fix unit test for sonic-config-engine.

**Comment by UmaMaven** [comment]
@qiluo-msft Please approve

File: src/sonic-yang-models/yang-models/sonic-static-route.yang [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+  revision 2022-03-17 {
+    description
+      "First Revision";
+  }
+
+  container sonic-static-route {
+    container STATIC_ROUTE {
+      description
+        "STATIC_ROUTE part of config_db.json";	
+	list STATIC_ROUTE_TEMPLATE_LIST {
```
Comment (qiluo-msft) [inline src/sonic-yang-models/yang-models/sonic-static-route.yang:32]: >	 [](http://example.com/codeflow?start=0&length=1)

Wrong indentation due to mixing tabs with spaces.

File: src/sonic-yang-models/yang-models/sonic-static-route.yang [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+            "prefix is the destination IP address, as key";
+        }
+        leaf nexthop {
+          type string;
+          description
+            "The next-hop that is to be used for the
+             static route as IP address. When interface needs to be
+             specified, use 0.0.0.0 as leaf value";
+        }
+        leaf ifname {
```
Comment (ganglyu) [inline src/sonic-yang-models/yang-models/sonic-static-route.yang:46]: can you use leafref?

File: src/sonic-yang-models/yang-models/sonic-static-route.yang [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+          type string;
+          description
+            "When interface is specified, forwarding happens through it";
+        }
+        leaf advertise {
+          type string{
+            pattern "((true|false),)*(true|false)";
+                  }
+          default "false";
```
Comment (ganglyu) [inline src/sonic-yang-models/yang-models/sonic-static-route.yang:55]: can you add unit test for this field?

File: src/sonic-yang-models/yang-models/sonic-static-route.yang [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            "prefix is the destination IP address, as key";
+        }
+        leaf nexthop {
+          type string;
+          description
+            "The next-hop that is to be used for the
+             static route as IP address. When interface needs to be
+             specified, use 0.0.0.0 as leaf value";
+        }
+        leaf ifname {
```
Comment (ganglyu) [inline src/sonic-yang-models/yang-models/sonic-static-route.yang:87]: leafref?

File: src/sonic-yang-models/tests/yang_model_tests/tests_config/static-route.json [hunk 5]
```diff
@@ -0,0 +1,1028 @@
...
+        },
+        "sonic-static-route:sonic-static-route": {
+            "sonic-static-route:STATIC_ROUTE": {
+                "STATIC_ROUTE_LIST": [{
+                        "prefix": "16.16.16.0/24",
+                        "vrf_name": "default",
+                        "nexthop": "1.0.0.4,2.0.0.3",
+                        "distance": "1,1",
+                        "nexthop-vrf": "vrf123,xyz",
+                        "blackhell": "false,false"
```
Comment (ganglyu) [inline src/sonic-yang-models/tests/yang_model_tests/tests_config/static-route.json:1020]: typo?

File: src/sonic-yang-models/yang-models/sonic-static-route.yang [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        leaf nexthop-vrf {
+          type string {
+            pattern "(((Vrf[a-zA-Z0-9_-]+)|(default)|(mgmt)),)*((Vrf[a-zA-Z0-9_-]+)|(default)|(mgmt))";
+          }
+          description
+            "VRF name of the nexthop. This is for vrf leaking";
+        }
+        leaf blackhole {
+          type string{
+            pattern "((true|false),)*(true|false)";
```
Comment (ganglyu) [inline src/sonic-yang-models/yang-models/sonic-static-route.yang:115]: How many boolean do you need?
Do you have any length limitation?

File: src/sonic-yang-models/yang-models/sonic-static-route.yang (hunk 2 above)
Comment (UmaMaven) [reply src/sonic-yang-models/yang-models/sonic-static-route.yang:46]: Leafref cannot be used as these are comma separated values to support ECMP.

File: src/sonic-yang-models/yang-models/sonic-static-route.yang (hunk 3 above)
Comment (UmaMaven) [reply src/sonic-yang-models/yang-models/sonic-static-route.yang:55]: Unit test has been added for STATIC_ROUTE_LIST.  STATIC_ROUTE_TEMPLATE_LIST was added to help with bgp unit tests. The STATIC_ROUTE schema in the community has only the prefix (no vrf -name as key).

File: src/sonic-yang-models/yang-models/sonic-static-route.yang (hunk 4 above)
Comment (UmaMaven) [reply src/sonic-yang-models/yang-models/sonic-static-route.yang:87]: Leafref cannot be used as these are comma separated values to support ECMP.

File: src/sonic-yang-models/yang-models/sonic-static-route.yang (hunk 6 above)
Comment (UmaMaven) [reply src/sonic-yang-models/yang-models/sonic-static-route.yang:115]: These are comma separated values. We can use multiple comma separated values with the last value without comma. The regex mandates that pattern.

File: src/sonic-yang-models/yang-models/sonic-static-route.yang (hunk 2 above)
Comment (ganglyu) [reply src/sonic-yang-models/yang-models/sonic-static-route.yang:46]: https://github.com/sonic-net/sonic-buildimage/blob/master/src/sonic-yang-models/yang-models/sonic-ntp.yang#L52-L79

Can you clarify the difference between comma separated interface and leaf-list?

File: src/sonic-yang-models/yang-models/sonic-static-route.yang (hunk 6 above)
Comment (ganglyu) [reply src/sonic-yang-models/yang-models/sonic-static-route.yang:115]: Do you have max ECMP size? I don't think we can support unlimited length string.

File: src/sonic-yang-models/yang-models/sonic-static-route.yang (hunk 2 above)
Comment (UmaMaven) [reply src/sonic-yang-models/yang-models/sonic-static-route.yang:46]: https://github.com/sonic-net/sonic-buildimage/pull/11499 has the same changes (got committed without the files during pulling to latest in github). 
Comma separated was added as part this PR, which would be of the format - interfaces will be "Ethernet0, Ethernet4"

File: src/sonic-yang-models/yang-models/sonic-static-route.yang (hunk 6 above)
Comment (UmaMaven) [reply src/sonic-yang-models/yang-models/sonic-static-route.yang:115]: As part of https://github.com/sonic-net/sonic-buildimage/pull/11499 we decided we have custom validation subsequently as a new PR. Please take a look at the comments there. Thanks

**Comment by UmaMaven** [comment]
Does anyone have anymore comments?
All the comments till now are addressed. 
Once you confirm, I will do the final code submission (after conflict resolution).

**Pull request closed without merging**